specification/testing_guidelines.md

```diff
@@ -0,0 +1,160 @@
+Testing guidelines
+==================
+
+These guidelines cover unit tests (being) developped at various parts of ```mini-kep``` project.
+
+We do not have functional (eg selenium) or load tests, which may have slightly different logic 
+or requirements in some aspects. 
+
+Introduction
+------------
+
+Testing is known to be even harder than writing original code. 
+
+When testing one must think of: 
+- how the code under test is supposed to run well 
+- how the code under test may run with error
+- how to expose behaviours above in test setup and checks?
+- how to write test in a clean and understandable fashion? how to name tests?
+- how to effectively mock functions to avoid third-party access and simulate real-time inputs 
+- what conditions / arguments are most probable for the code under test?
+- what happens in program outside code under test?
+
+Checklist
+----------
+
+Best tests:
+1. run quick and often
+2. use a continious integration like Travis CI and use coverage metrics like codecov 
+3. cover at least all public methods/functions
+4. have long names with context and expected result (except very short tests)
+5. are as simple and readable as they can get 
+6. one test tests one thing and has one assert per test
+7. have clear separation of setup, call of code under test and check 
+8. make good use of parametrisation, fixtures, dependency injection, mocks and monkey-patching
+9. concentrate around practical, not fantasy situations  
+10. include just a few integration, end-to-end tests
+11. fail early and close to where problem is
+12. show where are bugs in the code
+13. show where original code needs enhancements to cover possible edge situations
+14. are well organized - tests for common function are grouped in one class, etc.
+
+Learning
+--------
+Basic testing in python :
+- <http://docs.python-guide.org/en/latest/writing/tests/>
+
+Takeaways from <https://pylonsproject.org/community-unit-testing-guidelines.html>:
+- tests should be as simple as possible
+- tests should be isolated
+- each test method should test Just One Thing
+
+References about:
+- testing is not the best technique to ensure code quality
+
+Sample Tests
```

Comment:
This is an example of particular type of test - it must have a sunheader and explaination.